cli/src/main/java/com/box/l10n/mojito/cli/command/PullCommand.java

```diff
@@ -213,24 +233,86 @@ public void execute() throws CommandException {
             directoriesIncludePatterns,
             directoriesExcludePatterns)) {
 
-      consoleWriter.a("Localizing: ").fg(Color.CYAN).a(sourceFileMatch.getSourcePath()).println();
+      if (isParallel) {
+        consoleWriter
+            .a("Sending localize request for: ")
+            .fg(Color.CYAN)
+            .a(sourceFileMatch.getSourcePath())
+            .println();
+      } else {
+        consoleWriter.a("Localizing: ").fg(Color.CYAN).a(sourceFileMatch.getSourcePath()).println();
+      }
 
       List<String> filterOptions =
           commandHelper.getFilterOptionsOrDefaults(
               sourceFileMatch.getFileType(), filterOptionsParam);
 
-      if (localeMappingParam != null) {
-        generateLocalizedFilesWithLocaleMaping(repository, sourceFileMatch, filterOptions);
+      if (isParallel) {
+        sendContentForLocalizedGeneration(sourceFileMatch, filterOptions);
       } else {
-        generateLocalizedFilesWithoutLocaleMapping(repository, sourceFileMatch, filterOptions);
+        generateLocalizedFiles(sourceFileMatch, filterOptions);
       }
     }
 
+    if (isParallel) {
+      consoleWriter.println();
+      consoleWriter.a("Generating localized files").println();
+      pollForLocalizedFiles();
+    }
+
     writePullRunFileIfNeeded();
 
     consoleWriter.fg(Color.GREEN).newLine().a("Finished").println(2);
   }
 
+  private void generateLocalizedFiles(FileMatch sourceFileMatch, List<String> filterOptions) {
+    if (localeMappingParam != null) {
+      generateLocalizedFilesWithLocaleMaping(repository, sourceFileMatch, filterOptions);
+    } else {
+      generateLocalizedFilesWithoutLocaleMapping(repository, sourceFileMatch, filterOptions);
+    }
+  }
+
+  private void sendContentForLocalizedGeneration(
+      FileMatch sourceFileMatch, List<String> filterOptions) {
+    if (localeMappingParam != null) {
+      pollableTaskIdToFileMatchMap.put(
+          generateLocalizedFilesWithLocaleMappingParallel(
+                  repository, sourceFileMatch, filterOptions)
+              .getId(),
+          sourceFileMatch);
+    } else {
+      pollableTaskIdToFileMatchMap.put(
+          generateLocalizedFilesWithoutLocaleMappingParallel(
+                  repository, sourceFileMatch, filterOptions)
+              .getId(),
+          sourceFileMatch);
+    }
+  }
+
+  private void pollForLocalizedFiles() {
+    pollableTaskIdToFileMatchMap
+        .entrySet()
+        .parallelStream()
+        .forEach(
+            entry -> {
+              commandHelper.waitForPollableTaskSilencedOutput(entry.getKey());
```

Comment:
Would there be any concerns with having many parallel tasks that (eventually) call waitForPollableTask which, essentially, 99% of their time, do Thread.Sleep()? 

I'm assuming that for our volumes there isn't a practical concern that would warrant doing this with a Executors.newScheduledThreadPool().scheduleAtFixedRate(..) pattern instead to avoid needing to always have the Threads allocated - wanted to check just in case.

Reply:
In practice the `parallelStream` uses numOfCPUs - 1 threads so we'll only ever have that many active tasks running on the client at once. 

As the task itself is quite small (wait for server side task to finish, then write all localized assets to file) I don't think we require the extra overhead of using an Executor thread pool approach here.

Reply:
Looking from far away... this might not be optimal polling. Say the threads get stuck on the tasks that finish last, then the rest that could be download in the meantime is not and it is delaying the process.

It feel like we should have a queue of tasks to process, but if the task is not finished we move it to the end of the queue and start checking others. Multi-threading also seem good to have so I'd be looking into ExecutorService/ThreadPoolExecutors maybe? Thoughts?

Reply:
I guess my view is how big of an impact is that situation likely to have? As the waiting threads 'active' task (write files to disk) is quite small, and that we have multiple threads waiting, I don't expect much of an impact realistically on performance. 

If we have one thread polling the longest tasks I accept it could be writing while it's waiting but when the task does complete it'll iterate through the other completed tasks immediately. In the queued scenario there's a possibility that our polling just misses the top of queue's finish, moves that job to the back of the queue, same then happens for the next job etc. I'd be surprised if we're talking about more than a few seconds performance gain either way.

Reply:
> In practice the parallelStream uses numOfCPUs - 1 threads

Any thoughts on if we should potentially have more?

Reply:
As the majority of tasks done by the cli thread is i/o rather than cpu bound we probably could have more, my only minor concern would be contention around the lock on console write for writing to the console. However the recommendation for parallelStreams is to use the common pool rather than a custom pool, and the common pool is only increased via a global parameter which is why I left it as it was